index.bs

```diff
@@ -96,855 +84,643 @@ API which aims to do just that.
-   determines whether unknown HTML elements are to be considered. The default
-   is to drop them.
-   If this option is true, unkown HTML elements will still be checked against
-   all other built-in or configured checks.
-: allowComments
-:: The <dfn>allow comments option</dfn> determines whether HTML comments are
-   allowed.
-
-Note: `allowElements` creates a sanitizer that defaults to dropping elements,
-  while `blockElements` and `dropElements` defaults to keeping unknown
-  elements. Using both types is possible, but is probably of little practical
-  use. The same applies to `allowAttributes` and `dropAttributes`.
-
-<div class="example">
-```js
-  const sample = to_node("Some text <b><i>with</i></b> <blink>tags</blink>.");
-  const script_sample = to_node("abc <script>alert(1)</script> def");
-
-  // Some text <b>with</b> text tags.
-  new Sanitizer({allowElements: [ "b" ]}).sanitize(sample);
-
-  // Some text <i>with</i> <blink>tags</blink>.
-  new Sanitizer({blockElements: [ "b" ]}).sanitize(sample);
-
-  // Some text <blink>tags</blink>.
-  new Sanitizer({dropElements: [ "b" ]}).sanitize(sample);
-
-  // Note: The default configuration handles XSS-relevant input:
-
-  // Non-scripting input will be passed through:
-  new Sanitizer().sanitize(sample);  // Will output sample unmodified.
-
-  // Scripts will be blocked: "abc alert(1) def"
-  new Sanitizer().sanitize(script_sample);
-```
-</div>
-
-In addition to allow and block lists for elements and attributes, there are
-also options to configure some node or element types.
-
-Examples:
-```js
-  // Comments will be dropped by default.
-  const comment = to_node("Hello <!-- comment --> World!");
-  new Sanitizer().sanitize(comment);  // "Hello  World!"
-  new Sanitizer({allowComments: true}).sanitize(comment);  // Same as comment.
-```
+# Algorithms # {#algorithms}
 
-A sanitizer's configuration can be queried using the
-[=query the sanitizer config=] method.
+<div algorithm>
+To <dfn>set and filter HTML</dfn>, given an {{Element}} or {{DocumentFragment}}
+|target|, an {{Element}} |contextElement|, a [=string=] |html|, and a
+[=dictionary=] |options|, and a [=boolean=] flag |safe|, run these steps:
+
+1. If |safe| and |contextElement|'s [=Element/local name=] is "`script`" and
+   |contextElement|'s [=Element/namespace=] is the [=HTML namespace=] or the
+   [=SVG namespace=]:
+   1. Return.
+1. Let |config| be the result of calling [=canonicalize a configuration=] on
+   |options|["`sanitizer`"] and |safe|.
+1. Let |newChildren| be the result of the HTML [=fragment parsing algorithm=]
+   given |contextElement|, |html|, and `true`.
+1. Let |fragment| be a new {{DocumentFragment}} whose [=node document=] is |contextElement|'s [=node document=].
+1. [=list/iterate|For each=] |node| in |newChildren|, [=list/append=] |node| to |fragment|.
+1. If |config| exists:
+   1. Run [=sanitize=] on |fragment| using |config|.
```

Comment:
Inlined.

Config is conditional, because step 4 of canonicalize can return empty. The idea is that in the "unsafe" usages, one could in theory define a "no-op" config that doesn't do anything. But it seemed clearer to me to just skip the filtering step for "unsafe" ops where the user hasn't requested any filtering.

In terms of mechanism, I think the previous way was a bit confused. So I've now changed canonicalize to return an empty list / ordered map; and this check to check for emptyness.